### PR TITLE
[ci-visibility] Support `@playwright/test`

### LINKED ISSUE #1717: CI  Visibility: Support @playwright/test
Context: I'm discussing this DD product throughout: https://docs.datadoghq.com/continuous_integration/setup_tests/javascript/

This is a feature request to add first-class support for `@playwright/test` (https://github.com/microsoft/playwright) an amazing Microsoft project for end-to-end browser testing (including its own fabulous test runner).

Currently, DD and Playwright users can use the JUnit report option to take advantage of DD reporting and tools, but it would be nice to have a formal integration (at least for me, as a DD and Playwright user) like there is for Cypress/Mocha/Jest.

If this is something that the DD team would be willing to add to its set of existing plugins, I'd be happy to help contribute it. (Briefly looked at some of the other DD plugins which mostly make sense but would need a few pointers as to overall how they work on the DD side of things.)

Disclaimer: I'm only an **external** contributor to Playwright (https://github.com/microsoft/playwright/pulls?q=author%3Arwoll) and have no formal relationship with Microsoft.

### PR DESCRIPTION
**Pull request by juan-fernandez**
### What does this PR do?
Support [playwright](https://playwright.dev/) testing framework.

### Motivation
Closes #1717

### Plugin Checklist

- [ ] Unit tests.
- [x] TypeScript [definitions][1].
- [x] TypeScript [tests][2].
- [ ] API [documentation][3].
- [x] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/documentation/blob/master/content/en/tracing/trace_collection/library_config/nodejs.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### Additional Notes
Playwright does _not_ have a programmatic API (see discussion [here](https://github.com/microsoft/playwright/issues/7275)), so writing plugin tests for it becomes hard. The integration tests should cover all the logic though.

### CODE REVIEW

**Review by Qard — DISMISSED** [review]

**Comment by juan-fernandez** [comment]
thanks for the review @Qard . I just found an edge case testing an open source project so I'll need to do some additional changes. I've moved this to draft

**Comment by juan-fernandez** [comment]
it should be ready for re-review @Qard :)

File: integration-tests/ci-visibility-intake.js [hunk 1]
```diff
@@ -169,6 +169,22 @@ class FakeCiVisIntake extends FakeAgent {
     infoResponse = DEFAULT_INFO_RESPONSE
   }
 
+  gatherPayloads (payloadMatch, gatheringTime = 15000) {
+    const payloads = []
+    return new Promise((resolve, reject) => {
+      setTimeout(() => {
+        this.removeListener('message', messageHandler)
```
Comment (uurien) [inline integration-tests/ci-visibility-intake.js:176]: in the same method, can we use the same syntax? I mean: `removeListener` with `addListener` or `on` with `off`

File: integration-tests/ci-visibility/playwright-tests/todo-list-page-test.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+const { test, expect } = require('@playwright/test')
+
+test.beforeEach(async ({ page }) => {
+  await page.goto(process.env.PW_BASE_URL)
+})
+
+test.describe('playwright', () => {
+  test('should work with failing tests', async ({ page }) => {
+    await expect(page.locator('.hello-world')).toHaveText([
+      'Hello Warld'
```
Comment (uurien) [inline integration-tests/ci-visibility/playwright-tests/todo-list-page-test.js:10]: "Warld"? If this is really a typo... it looks like the test is not checking it correctly.

File: packages/datadog-plugin-playwright/src/index.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    this._testSuites = new Map()
+
+    this.addSub('ci:playwright:session:start', ({ command, frameworkVersion, rootDir }) => {
+      const childOf = getTestParentSpan(this.tracer)
+      this.command = command
+      this.frameworkVersion = frameworkVersion
+      this.rootDir = rootDir
+
+      const testSessionSpanMetadata = getTestSessionCommonTags(command, frameworkVersion)
+      this.testSessionSpan = this.tracer.startSpan('playwright.test_session', {
```
Comment (uurien) [inline packages/datadog-plugin-playwright/src/index.js:37]: this is saved as state in the Plugin... does it means that testSession is unique in whole execution?

File: integration-tests/ci-visibility/playwright-tests/todo-list-page-test.js (hunk 2 above)
Comment (uurien) [reply integration-tests/ci-visibility/playwright-tests/todo-list-page-test.js:10]: "should work with failing tests"
ok, is expected to fail, sorry

Comment (juan-fernandez) [reply integration-tests/ci-visibility/playwright-tests/todo-list-page-test.js:10]: indeed 😄

File: integration-tests/ci-visibility-intake.js (hunk 1 above)
Comment (juan-fernandez) [reply integration-tests/ci-visibility-intake.js:176]: yeah that's a good idea

File: packages/datadog-plugin-playwright/src/index.js (hunk 3 above)
Comment (juan-fernandez) [reply packages/datadog-plugin-playwright/src/index.js:37]: that's a good question: yes, there's a single test session for the whole execution. There's parallelism in playwright, but it happens at a lower level 😄

**Comment by juan-fernandez** [comment]
funnily enough there was a playwright release just yesterday and it seems it's not compatible with this instrumentation 😅 . I'll check if it's an easy fix

**Comment by juan-fernandez** [comment]
> funnily enough there was a playwright release just yesterday and it seems it's not compatible with this instrumentation 😅 . I'll check if it's an easy fix

ok that wasn't too bad but it required yet another small rewrite 😆 . Ready for review finally 🤞

File: packages/datadog-plugin-playwright/src/index.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+    this.addSub('ci:playwright:test:start', ({ testName, testSuiteAbsolutePath }) => {
+      const store = storage.getStore()
+      const testSuite = getTestSuitePath(testSuiteAbsolutePath, this.rootDir)
+      const span = this.startTestSpan(testName, testSuite)
+
+      this.enter(span, store)
+    })
+    this.addSub('ci:playwright:test:finish', ({ testStatus, steps, error }) => {
+      const span = storage.getStore().span
```
Comment (Qard) [inline packages/datadog-plugin-playwright/src/index.js:91]: Probably shouldn't just assume a span was found. Context loss could happen, we should code in a way that avoids the process crashing when it does.

```suggestion
      const store = storage.getStore()
      const span = store && store.span
      if (!span) return
```

Might also be worth copying the `activeSpan` getter from TracingPlugin into CiPlugin or maybe even the Plugin base class if you're going to need this pattern a bunch.

**Review by Qard — APPROVED** [review]
LGTM with one nit.

File: packages/datadog-plugin-playwright/src/index.js (hunk 4 above)
Comment (juan-fernandez) [reply packages/datadog-plugin-playwright/src/index.js]: good idea, thanks. I've added it just here but I'll see if I can do the getter in another PR 👍

File: packages/datadog-instrumentations/src/playwright.js [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+addHook({
+  name: '@playwright/test',
+  file: 'lib/runner.js',
+  versions: ['>=1.18.0']
+}, runnerHook)
+
+addHook({
+  name: '@playwright/test',
+  file: 'lib/dispatcher.js',
+  versions: ['1.18.0 - 1.29.2']
```
Comment (uurien) [inline packages/datadog-instrumentations/src/playwright.js:229]: If they release a patch for `1.29.x` we are going to ignore it... what do you think about this version range?

```suggestion
  versions: ['>=1.18.0  <1.30.0']
```

Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/playwright.js]: good catch 👍

File: packages/datadog-instrumentations/src/playwright.js [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    const worker = createWorker.apply(this, arguments)
+
+    worker.on('testBegin', ({ testId }) => {
+      const { test } = dispatcher._testById.get(testId)
+      testBeginHandler(test)
+    })
+    worker.on('testEnd', ({ testId, status, errors }) => {
+      const { test } = dispatcher._testById.get(testId)
+
+      testEndHandler(test, STATUS_TO_TEST_STATUS[status], errors[0])
```
Comment (uurien) [inline packages/datadog-instrumentations/src/playwright.js:176]: Has `errors` a value when the test is success? Should we protect it to be sure that it works if they change the implementation in the future?
```suggestion
      testEndHandler(test, STATUS_TO_TEST_STATUS[status], errors && errors[0])
```

**Review by uurien — APPROVED** [review]

File: packages/datadog-instrumentations/src/playwright.js (hunk 6 above)
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/playwright.js:176]: yes, it should always be defined, at least for the versions we support right now. It's a good idea to check though, so I'll do it in a follow up PR

Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/playwright.js:176]: I'm working on https://github.com/DataDog/dd-trace-js/pull/2725 right now, which will include some playwright changes, so I'll add this suggestion there

**Pull request merged**